third_party/move/move-model/bytecode/abstract_domain_derive/tests/test.rs

```diff
@@ -0,0 +1,79 @@
+use abstract_domain_derive::AbstractDomain;
+use move_stackless_bytecode::dataflow_domains::{AbstractDomain, JoinResult};
+
+#[derive(AbstractDomain)]
+struct Unit;
+
+/// Add a formal top and bottom element to type `T`
+#[derive(PartialEq, Eq, Clone)]
+pub enum Plus2<T> {
+    Top,
+    Mid(T),
+    Bot,
+}
+
+impl<T: Eq + Clone> AbstractDomain for Plus2<T> {
+    fn join(&mut self, other: &Self) -> JoinResult {
+        match (&self, other) {
+            (Plus2::Top, _) | (_, Plus2::Bot) => JoinResult::Unchanged,
+            (Plus2::Mid(_), _ /* Mid or Top */) => {
```

Comment:
As I mentioned in a different version of this code, I think this is losing precision.

Consider `Mid(x)` joined with itself. This will take it to `Top` unnecessarily losing precision.

Separately, if `T` is itself an `AbstractDomain`, we could use joins on `T` to improve precision when joining mids.

Reply:
I realized this is test code, but because of the tendency to copy around such code or refer to them, would be nice to either fix this or have a clear comment about loss of precision.

Reply:
done. (If `T` is itself an `AbstractDomain`, then we won't need to wrap it with `Plus2` since it already has a top (and bottom?) element. The intended use of `Plus2` is to construct a lattice out of a set with no order structure.)